AP_Scripting: add binding and example for set_origin()

CallanDaniel opened the pull request:
Added binding for setting EKF origin, now that this fix is in [https://github.com/ArduPilot/ardupilot/pull/17786](https://github.com/ArduPilot/ardupilot/pull/17786)

Tested in SITL with simulated vicon:
![set_origin console](https://user-images.githubusercontent.com/43185766/123977062-f86f7600-d9be-11eb-831a-169d14092357.JPG)

IamPete1 commented on libraries/AP_Scripting/examples/ahrs-set-origin.lua:
```diff
@@ -0,0 +1,22 @@
+
+    if origin then
+        gcs:send_text(0, "Unable to set EKF origin: already set")
+    return
```
since this is just a return you could do

```
origin = assert(ahrs:get_origin(),"Unable to set EKF origin: already set")

```
on line 8

hendjoshsr71 replied:
Also with the early return here you don't need the `else` afterwards.
L13: indent of the return is wrong

IamPete1 commented on libraries/AP_Scripting/examples/ahrs-set-origin.lua:
```diff
@@ -0,0 +1,22 @@
+
+function update ()
+
+    home = ahrs:get_home()
```
Your not using this

IamPete1 commented on libraries/AP_Scripting/examples/ahrs-set-origin.lua:
```diff
@@ -0,0 +1,22 @@
+        gcs:send_text(0, "Unable to set EKF origin: already set")
+    return
+    else
+        ahrs:set_origin(location)
```
you should check the return of this, make sure it has actually set.

hendjoshsr71 replied:
^^This: I still had gps enabled and it prints "Set" a bunch of times until it gets a lock and the origin is finally set.

hendjoshsr71 replied:
If you did check it you could do something like this and bring back your printing
```suggestion
    if ahrs:set_origin(location) then
        gcs:send_text(0, string.format("Origin Set - Lat:%.7f Long:%.7f Alt:%.1f", location:lat()/10000000, location:lng()/10000000, location:alt()/100))
    else
        gcs:send_text(0, string.format("Origin Refused to be Set"))
    end
```

CallanDaniel commented:
Really appreciate the reviews! @hendjoshsr71 @IamPete1 I'll work through them asap :)

CallanDaniel commented:
@rmackay9  @hendjoshsr71 I've changed the script, hopefully this is better. I also added a binding to check if the AHRS is initialised, to avoid getting a bunch of "EKF3 refusing to set origin"  warnings during initialisation while testing with GPS enabled.

In SITL with GPS disabled and simulated VICON it sets the origin and then exits the script with the error "EKF origin already set"
In SITL with GPS enabled (stock copter sitl params) it exits the script with the error "EKF origin already set" 

let me know if this is sufficent.

hendjoshsr71 commented on libraries/AP_Scripting/examples/ahrs-set-origin.lua:
```diff
@@ -0,0 +1,16 @@
+        ahrs:set_origin(location)
+    end
+
+    return update, 5000
```
Whether successful or not, after setting the origin we don't need to call the script back right?
```suggestion
    return
```

Prevents getting the error from the assert when we were successful

hendjoshsr71 commented on libraries/AP_Scripting/examples/ahrs-set-origin.lua:
```diff
@@ -0,0 +1,16 @@
+
+function update ()
+
+    if ahrs:initialised() then
```
Could do an early return
```suggestion
     if not ahrs:initialised() then
        return update, 5000
    end
```

hendjoshsr71 reviewed (changes requested):
@CallanDaniel  I like really the addition of `ahrs:initialised()` as it solves the problem I had when I first tried the assert the other day.

You dropped printing of where the origin is set to which is useful, IMO. 
Especially, since even on success you get this message.
![image](https://user-images.githubusercontent.com/69225461/124335729-4cb65980-db69-11eb-9ddc-ea02aa715876.png)

BTW I'm learning Lua too so my suggestions are what I came up with based upon your work, but more experienced people may have better ideas :)

- It restructures it to use an early return. (fewer indents)
- Checks the return of `ahrs:set_origin`
- Prints out success
- returns such that we don't get an error message from the assert, even if set was successful

CallanDaniel commented:
@hendjoshsr71 Thanks, this is really all a learning exercise for me so I really do appreciate the time and input! I thought removing the printing might make everything simpler, but I totally see where you are coming from now. I also see what you mean checking the return of ahrs:set_origin(location) to confirm it's actually been set. The early return is also a good idea. Apologies if it takes me a while to reply, timezones, work and such...

Thanks again for the help

CallanDaniel commented:
@hendjoshsr71 okay, so I've updated the script as you've suggested, but now it seems to me as if 
```
origin = assert(not ahrs:get_origin(),"Refused to set EKF origin - already set")
```
and
```
else
        gcs:send_text(0, "Refused to set EKF origin")
end
```

are kind of doing the same thing? 

In GPS enabled modes where setting origin is not aloud the script ends with an error on the assert satatement so it never really get to the "else" branch of the if statement.

hendjoshsr71 commented:
> @hendjoshsr71 okay, so I've updated the script as you've suggested, but now it seems to me as if
> 
> ```
> origin = assert(not ahrs:get_origin(),"Refused to set EKF origin - already set")
> ```
> 
> and
> 
> ```
> else
>         gcs:send_text(0, "Refused to set EKF origin")
> end
> ```
> 
> are kind of doing the same thing?
> 
> In GPS enabled modes where setting origin is not aloud the script ends with an error on the assert satatement so it never really get to the "else" branch of the if statement.

@CallanDaniel 

In a GPS enabled mode: This script can try to set the origin before GPS lock is achieved, but the request is refused. The origin has not been set, but it can get set later when a lock is achieved. Thus you will pass the assert, but `ahrs:set_origin(location)` will pass false and we wouldn't know it otherwise. 

When I originally put in the `print` in I didn't remember about that case. But what happens if later on code gets added or removed in dependent functions? The conditions for passing back false have changed, but we don't catch it here which could user annoyance.  

This is sort of like defensive coding which you will see for important items like the origin_set issue the original PR exposed. Now we check at every dependent calling function `if (valid_origin)`. https://github.com/ArduPilot/ardupilot/pull/17786  Is that necessary? No. We could just ensure that the original calling function does that. 

Really though this is an example script so it makes sense to demonstrate the bindings you've added: inputs & outputs. Even if the example seems trivial. 

Also notice what the change from `return update, 5000` to `return` does. We will never call this function back again thus the assert can't catch it and print out to let the user know `set_origin` was refused.

CallanDaniel commented:
@hendjoshsr71 Ah okay, that makes sense, thanks!

hendjoshsr71 commented:
@CallanDaniel Are you able to/desire to attend the devcall on monday? I think the Tuesday call works better for your timezone?

rmackay9 commented:
I think we shouldn't worry too much about making the lua script perfect.  These scripts are mostly just to provide an example of how the method can be called.  If the EKF origin can be set then I think it is OK to merge... and I might do that if there are no objections..

hendjoshsr71 commented:
It worked well for me in SITL, so I think it is good to go.

rmackay9 commented:
Merged, thanks!
